### PR TITLE
Reindex conversions

### PR DESCRIPTION
**Pull request by jesperrojestal**
Small module w/ documentation to just reindex the Searchjoy conversions when used together with Searchkick. Rough first proof of concept but tested in our production environment and speeds up reindexing conversions from more than 1 hour down to 3 minutes.

### CODE REVIEW

File: lib/searchjoy/conversions.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        end
+      end
+
+      reindex_map.keys.each do |klass_name|
+        klass = klass_name.constantize
+
+        batch_size ||= klass.searchkick_options[:batch_size] if klass.respond_to?(:searchkick_options)
+        batch_size ||= 1_000
+
+        Searchkick.callbacks(callback) do
```
Comment (ankane) [inline lib/searchjoy/conversions.rb]: This should go inside `find_in_batches` - otherwise, it'll queue up all the reindexes and do them once at the end.

**Comment by ankane** [comment]
Hi Jesper, thanks for the PR!  In a similar fashion, we only reindex items that have a recent conversion (for instance, if we were to reindex every 4 hours, we'd look for conversions within 4 hours).

A few thoughts:

- Not sure we want to create a `Searchjoy::Conversions` class for this. We could instead do:
  ```ruby
  Searchjoy.reindex_conversions(within: 4.hours)
  ```
- May be nice to have an option to reindex specific `convertable_type`s
- Does it make sense to support callbacks other than `bulk`?
- I'm not sure we want to have options to modify the logger, but if so, we'll need to set it back to the original afterwards or in case of an error (with an `ensure` block)
- Please commit only the minimum changeset to make this work (no style changes, `freeze`s, etc)

Let me know what you think.

**Comment by jesperrojestal** [comment]
Hello Andrew!

- Searchjoy::Conversions was chosen because:
1. I created the PR from my original Rails module i wrote in the beginning, so it was easiest to just move most of it into a submodule of Searchjoy due to the simple fact that i'm lazy. ;)
2. I didn't want to pollute the Searchjoy namespace when submitting new features.
3. Conversions reindex relies on Searchkick but is a modification to Searchjoy. So where to place it? My first thought was a submodule. But thats more of a code-philosophical problem. :)
But if you're fine with it being in the root Searchjoy namespace, i'll add it to the list of TODOs.

- :convertable_type(s?) and :within is a excellent idea but maybe use :from as option key, like:
```ruby
Searchjoy.reindex_conversions(from: 4.hours.ago)
```
it would also be easier for non-english speaking users to understand what it does without writing too much documentation, just my thoughts

- Callbacks other than :bulk might good for tests in the future? was fiddling around with the idea to let each async worker reindex each batch find_in_batches to have concurrent reindexing and speed it up even further. BTW,  Searchkick.callbacks could use better documentation, :update_fields method?

- Logger is on the TODOs

- Sorry for the commit bloat, i have a tendency to fix what rubocop code faschist whines about as i go along.

Let me know your thoughts.

**Comment by jesperrojestal** [comment]
Feels like i can't improve on this much more now... any suggestions?

**Comment by ACPK** [comment]
@jesperrojestal - Is this moving forward in your opinion?

**Comment by jesperrojestal** [comment]
@ACPK - this is more or less ready to merge for a while now and no suggestions. Been running in our production env for a while now and performs extremely well.

**Comment by ankane** [comment]
Hey @jesperrojestal, I documented the approach I'd recommend for speeding up conversions here: https://github.com/ankane/searchkick#performant-conversions

I appreciate the PR.

**Pull request closed without merging**